torch/_decomp/decompositions.py

```diff
@@ -3273,64 +3287,82 @@ def upsample_bilinear2d(
-    else:
-        h_scale_factor = 0.0
+            scale_factor = 0.0
 
-    if out_w > 1:
+        i = torch.arange(out, dtype=input.dtype, device=input.device)
         if align_corners:
-            w_scale_factor = (in_w - 1) / (out_w - 1)
+            x = scale_factor * i
         else:
-            w_scale_factor = 1.0 / scales_w if scales_w is not None else in_w / out_w
-    else:
-        w_scale_factor = 0.0
-
-    i = torch.arange(out_h, dtype=input.dtype, device=input.device)
-    j = torch.arange(out_w, dtype=input.dtype, device=input.device)
-
-    if align_corners:
-        x = h_scale_factor * i
-        y = w_scale_factor * j
-    else:
-        x = (h_scale_factor * (i + 0.5) - 0.5).clamp(min=0.0)
-        y = (w_scale_factor * (j + 0.5) - 0.5).clamp(min=0.0)
-
-    x_floor = x.to(torch.int64)
-    x_ceil = torch.ceil(x).clamp(max=in_h - 1).to(torch.int64)
-    y_floor = y.to(torch.int64)
-    y_ceil = torch.ceil(y).clamp(max=in_w - 1).to(torch.int64)
-
-    x_view = x.unsqueeze(1)
-    x_floor_view = x_floor.unsqueeze(1)
-    x_ceil_view = x_ceil.unsqueeze(1)
-
-    v1 = aten._unsafe_index(input, [None, None, x_floor_view, y_floor])
-    v2 = aten._unsafe_index(input, [None, None, x_ceil_view, y_floor])
-    v3 = aten._unsafe_index(input, [None, None, x_floor_view, y_ceil])
-    v4 = aten._unsafe_index(input, [None, None, x_ceil_view, y_ceil])
-
-    xscale2 = x_view - x_floor_view
-    xscale1 = 1.0 - xscale2
-
-    yscale2 = y - y_floor
-    yscale1 = 1.0 - yscale2
-
-    q1 = torch.mul(v1, xscale1) + torch.mul(v2, xscale2)
-    q2 = torch.mul(v3, xscale1) + torch.mul(v4, xscale2)
-    result = torch.mul(q1, yscale1) + torch.mul(q2, yscale2)
+            x = (scale_factor * (i + 0.5) - 0.5).clamp(min=0.0)
+        x_floor = x.to(torch.int64)
+        x_ceil = torch.ceil(x).clamp(max=inp - 1).to(torch.int64)
+        return x, x_floor, x_ceil
+
+    x, x_floor, x_ceil = get_values(in_d, out_d, scales_d)
+    y, y_floor, y_ceil = get_values(in_h, out_h, scales_h)
+    z, z_floor, z_ceil = get_values(in_w, out_w, scales_w)
+
+    views = [x.unsqueeze(1).unsqueeze(2), y.unsqueeze(1), z]
+    floor_views = [x_floor.unsqueeze(1).unsqueeze(2), y_floor.unsqueeze(1), z_floor]
+    ceil_views = [x_ceil.unsqueeze(1).unsqueeze(2), y_ceil.unsqueeze(1), z_ceil]
+
+    result = 0
+    d = 3
+    for j in range(2**d):
+        # we write j in binary format to see if we are using
+        # ceil or floor.
+        n = format(j, f"00{d}b")
+        idx = [None, None]
+        scales = []
+        for k in range(d):
+            if n[k] == "0":
+                idx.append(floor_views[k])
+                scales.append(1.0 - (views[k] - floor_views[k]))
+            else:
+                idx.append(ceil_views[k])
+                scales.append(views[k] - floor_views[k])
+        v = aten._unsafe_index(input, idx)
+        for scale in scales:
+            v = torch.mul(v, scale)
+        result += v
```

Comment:
No in-place ops please, as these need to be removed by the compiler afterwards. Do `result = result + v`